### PR TITLE
Omen models. First attempt.

### PR DESCRIPTION
**Pull request by hdser**
# Thank you for contributing to Spellbook!

Thank you for taking the time to submit code in Spellbook. A few things to consider:

- If you are a first-time contributor, please sign the CLA by copy & pasting **_exactly_** what the bot mentions in PR comment
- Refer to [docs](#spellbook-contribution-docs) section below to answer questions
- Dune team will review submitted PRs as soon as possible

### To speed up your development process in PRs, keep these tips in mind:

- Each commit to your feature branch will rerun CI tests ([see example](https://github.com/duneanalytics/spellbook/actions/runs/8202519819/job/22433451880?pr=5519))
    - This includes *all* modified models on your branch
    - This includes *all* history of the data
- Two tips for faster development iteration:
    - Ensure dbt is installed locally (refer to main `readme`) and run `dbt compile`
        - This will output raw SQL in `target/` directory to copy/paste and run on Dune directly for initial query testing
    - Hardcode a `WHERE` filter for only ~7 days of history on large source tables, i.e. `ethereum.transactions`
        - This will speed up the CI tests and output results quicker -- whether that's an error or fully successful run
        - Once comfortable with small timeframe, remove filter and let full history run

### Incremental model setup
- Make sure your unique key columns are *exactly* the same in the model config block, schema yml file, and seed match columns (where applicable)
- There cannot be nulls in the unique key columns
    - Be sure to double check key columns are correct or `COALESCE()` as needed on key column(s), otherwise the tests may fail on duplicates

### 🪄 Use the built CI tables for testing 🪄

Once CI completes, you can query the CI tables and errors in dune when it finishes running.
- For example:
    - In the `run initial models` and `test initial models`, there will be a schema that looks like this: `test_schema.git_dunesql_4da8bae_sudoswap_v2_base_pool_creations`
    - This can be temporarily queried in Dune for ~24 hours

Leverage these tables to perform QA testing on Dune query editor -- or even full test dashboards!

## Spellbook contribution docs

The [docs](docs) directory has been implemented to answer as many questions as possible. Please take the time to reference each `.md` file within this directory to understand how to efficiently contribute & why the repo is designed as it is 🪄

Example questions to be answered:

- What does each property in the [model config block](/docs/models/model_config_block.md) mean?
- What is the [CI test](/docs/ci_test/ci_test_overview.md) attached to PRs and how can I best utilize it?
- What are the Spellbook [best practices](/docs/general/best_practices.md)?

Please navigate through the [docs](/docs) directory to find as much info as you can.

**Note**: happy to take PRs to improve the docs, let us know 🤝

### CODE REVIEW

**Comment by hdser** [comment]
I have read the CLA Document and I hereby sign the CLA

File: models/omen/gnosis/omen_gnosis_markets.sql [hunk 1]
```diff
@@ -0,0 +1,185 @@
+{{ config(
+    schema = 'omen_gnosis',
+    alias = 'markets',
+    
+    partition_by = ['block_day'],
+    materialized = 'incremental',
+    file_format = 'delta',
+    incremental_strategy = 'merge',
+    unique_key = ['fixedProductMarketMaker', 'questionId', 'conditionId', 'conditionIds_index'],
```
Comment (Hosuke) [inline models/omen/gnosis/omen_gnosis_markets.sql:9]: `conditionIds_index` is not existed in the final selection in this model.

File: models/omen/gnosis/omen_gnosis_liquidity.sql [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+            VARBINARY_TO_UINT256(VARBINARY_LTRIM(VARBINARY_SUBSTRING(data, 33, 32))) AS sharesMinted,
+            VARBINARY_TO_UINT256(VARBINARY_LTRIM(VARBINARY_SUBSTRING(data, 65, 32))) AS amountsAdded_size,
+            data
+        FROM 
+            {{source('gnosis','logs') }}
+        WHERE
+            --FPMMDeterministicFactory_evt_FPMMFundingAdded
+            topic0 = 0xec2dc3e5a3bb9aa0a1deb905d2bd23640d07f107e6ceb484024501aad964a951 
+            {% if is_incremental() %}
+            AND block_time >= date_trunc('day', now() - interval '7' day)
```
Comment (Hosuke) [inline models/omen/gnosis/omen_gnosis_liquidity.sql:57]: We can use `incremental_predicate` here and for other incremental filters:
```suggestion
            AND {{ incremental_predicate('block_time') }}
```

Also if you think this PR is ready-for-review, please mark it ready from draft status.

File: models/omen/gnosis/omen_gnosis_liquidity.sql [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    
+    partition_by = ['block_day'],
+    materialized = 'incremental',
+    file_format = 'delta',
+    incremental_strategy = 'merge',
+    unique_key = ['block_time', 'tx_hash', 'evt_index'],
+    post_hook='{{ expose_spells(\'["gnosis"]\',
+                                "project",
+                                "omen",
+                                \'["hdser"]\') }}'
```
Comment (Hosuke) [inline models/omen/gnosis/omen_gnosis_liquidity.sql:13]: We may use a better format for the post_hook here (and in future):
```suggestion
    post_hook='{{ expose_spells(blockchains  = \'["gnosis"]\',
                                spell_type   = "project",
                                spell_name   = "omen",
                                contributors = \'["hdser"]\') }}'
```

File: models/omen/gnosis/omen_gnosis_liquidity.sql [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        ) AS outcomeTokens_amount
+    FROM 
+        {{source('gnosis','logs') }}
+    WHERE
+        --FPMMDeterministicFactory_evt_FPMMFundingAdded
+        topic0 = 0xec2dc3e5a3bb9aa0a1deb905d2bd23640d07f107e6ceb484024501aad964a951 
+        {% if is_incremental() %}
+        AND block_time >= date_trunc('day', now() - interval '7' day)
+        AND 
+        {{ incremental_predicate('block_time') }}
```
Comment (Hosuke) [inline models/omen/gnosis/omen_gnosis_liquidity.sql:49]: These 2 lines are actually equivalent.
```suggestion
        AND {{ incremental_predicate('block_time') }}
```

File: models/omen/gnosis/omen_gnosis_liquidity.sql [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    FROM 
+        {{source('gnosis','logs') }}
+    WHERE
+        --FPMMDeterministicFactory_evt_FPMMFundingRemoved
+        topic0 = 0x8b4b2c8ebd04c47fc8bce136a85df9b93fcb1f47c8aa296457d4391519d190e7
+        {% if is_incremental() %}
+        AND 
+        block_time >= date_trunc('day', now() - interval '7' day)
+        AND 
+        {{ incremental_predicate('block_time') }}
```
Comment (Hosuke) [inline models/omen/gnosis/omen_gnosis_liquidity.sql:107]: Same as above.
```suggestion
        AND {{ incremental_predicate('block_time') }}
```

File: models/omen/gnosis/omen_gnosis_liquidity.sql [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        FPMMFundingRemoved 
+),
+
+final AS (
+    SELECT * FROM add_liquidity
+    UNION ALL
+    SELECT * FROM remove_liquidity
+)
+
+SELECt * FROM final
```
Comment (Hosuke) [inline models/omen/gnosis/omen_gnosis_liquidity.sql:140]: We may list all columns here instead of using `*`(for matching with the schema.yml).

**Review by Hosuke — CHANGES_REQUESTED** [review]

File: models/omen/gnosis/omen_gnosis_liquidity.sql (hunk 6 above)
Comment (hdser) [reply models/omen/gnosis/omen_gnosis_liquidity.sql]: Updated.

**Review by Hosuke — APPROVED** [review]
LGTM.✅
Thank you @hdser 

Great use of `TRANSFORM`.

[Sample omen_gnosis_trades]()
[Sample omen_gnosis_liquidity]()

File: models/omen/gnosis/omen_gnosis_liquidity.sql [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    tx_hash,
+    evt_index,
+    fixedProductMarketMaker,
+    funder,
+    shares,
+    collateralRemovedFromFeePool,
+    outcomeIndex,
+    outcomeTokens_amount,
+    action,
+    reserves_delta
```
Comment (jeff-dude) [inline models/omen/gnosis/omen_gnosis_liquidity.sql:149]: can we stick to one standard on column names?
preference in spellbook is lowercase, underscores

Comment (jeff-dude) [reply models/omen/gnosis/omen_gnosis_liquidity.sql:149]: make sure to update schema yml file to account for column name changes

File: models/omen/gnosis/omen_gnosis_trades.sql [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    ,evt_index
+    ,fixedProductMarketMaker
+    ,address
+    ,outcomeIndex
+    ,outcomeSlot
+    ,amount
+    ,feeAmount
+    ,outcomeTokens_amount
+    ,action
+    ,reserves_delta
```
Comment (jeff-dude) [inline models/omen/gnosis/omen_gnosis_trades.sql:122]: same comment here on column format

File: models/omen/gnosis/omen_gnosis_trades.sql [hunk 9]
```diff
@@ -0,0 +1,123 @@
+{{ config(
+    schema = 'omen_gnosis',
+    alias = 'trades',
+    
+    partition_by = ['block_day'],
```
Comment (jeff-dude) [inline models/omen/gnosis/omen_gnosis_trades.sql:5]: i would recommend removing the partition column altogether. we don't want to over-partition, or it'll make performance worse. you can see how small each day is:
```
select
  block_day,
  count(1)
from
  test_schema.git_dunesql_0191ef6_omen_gnosis_trades
group by
  block_day
```

Comment (jeff-dude) [reply models/omen/gnosis/omen_gnosis_trades.sql:5]: this likely applies to both models

File: models/omen/gnosis/omen_gnosis_trades.sql [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+}}
+
+{% set project_start_date = '2020-12-01' %}
+
+WITH
+
+trades AS (
+    SELECT
+        block_time
+        ,DATE_TRUNC('day', block_time) AS block_day
```
Comment (jeff-dude) [inline models/omen/gnosis/omen_gnosis_trades.sql:24]: i'd recommend explicitly casting to date, otherwise it'll default to `block_time` timestamp value

**Review by jeff-dude — CHANGES_REQUESTED** [review]

**Review by jeff-dude — APPROVED** [review]

**Pull request merged**